### PR TITLE
Store items out-of-line in the HIR

### PR DESCRIPTION
**Pull request by nikomatsakis**
This PR moves items into a separate map stored in the krate, rather than storing them inline in the HIR. The HIR visitor is also modified to skip visiting nested items by default. The goal here is to ensure that if you get access to the HIR for one item, you don't automatically get access to a bunch of other items, for better dependency tracking.

r? @nrc 
cc @eddyb

### CODE REVIEW

File: src/librustc_front/hir.rs [hunk 1]
```diff
@@ -320,13 +322,42 @@ pub struct WhereEqPredicate {
...
+    /// approach. You should override `visit_nested_item` in your
+    /// visitor and then call `intravisit::walk_crate` instead.
+    pub fn visit_all_items<'hir, V:Visitor<'hir>>(&'hir self, visitor: &mut V) {
+        // In principle, we could just iterate over the hashmap, but
+        // in practice that makes the order of error reporting vary
+        // with small changes in the input etc etc, which makes the
+        // test base hard to maintain. So instead we sort by node-id
+        // so as to get reproducible results.
+        let mut pairs: Vec<_> = self.items.iter().collect();
+        pairs.sort_by(|&(id1, _), &(id2, _)| id1.cmp(id2));
```
Comment (eddyb) [inline src/librustc_front/hir.rs]: Couldn't this be pre-computed? Or use `BTreeMap` instead.
Or compact Item IDs instead of Node IDs.

**Comment by nrc** [comment]
r+ with the comment (this is surprisingly straightforward, nice).

File: src/librustc_front/hir.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc_front/hir.rs]: @eddyb yeah, I was debating about this. A `BtreeMap` would certainly be easy enough to do for now. Compact item-ids seems like the best solution all around perhaps, but I feel like I'd prefer to do that sort of change as a separate patch.

File: src/librustc_trans/trans/base.rs [hunk 2]
```diff
@@ -2315,8 +2309,9 @@ pub fn trans_item(ccx: &CrateContext, item: &hir::Item) {
                          generics,
                          item.id);
       }
-      hir::ItemMod(ref m) => {
-        trans_mod(&ccx.rotate(), m);
+      hir::ItemMod(_) => {
+          // modules have no equivalent at runtime, they just affect
+          // the mangled names of things contained within
```
Comment (sanxiyn) [inline src/librustc_trans/trans/base.rs]: Doesn't this make `-C codegen-units` ineffective?

Comment (nikomatsakis) [inline src/librustc_trans/trans/base.rs]: @sanxiyn oh, good catch! I totally overlooked that, actually!

**Comment by nikomatsakis** [comment]
I had @michaelwoerister review last commit.

**Comment by nikomatsakis** [comment]
@bors r=nrc,mw

Hopefully this will pass. I've been able to run make check locally to completion due to various local problems.
